cpp/validate_harvested_records.cc

```diff
@@ -322,126 +244,122 @@ void SendEmail(const std::string &email_address, const std::string &message_subj
-    if (journal_name_and_info == journal_name_to_info_map->end()) {
-        first_record = true;
-        JournalInfo new_journal_info;
-        LoadFromDatabaseOrCreateFromScratch(db_connection, zeder_id, zeder_instance, &new_journal_info);
-        (*journal_name_to_info_map)[journal_name] = new_journal_info;
-        journal_name_and_info = journal_name_to_info_map->find(journal_name);
+    reasons_for_being_invalid->clear();
+
+    const auto zid_field(record.findTag("ZID"));
+    if (unlikely(zid_field == record.end()))
+        LOG_ERROR("record is missing a ZID field!");
+    else if (zid_field->getFirstSubfieldWithCode('a').empty())
+        LOG_ERROR("record is missing an a-subfield in the existing ZID field!");
+    const auto ZID(zid_field->getFirstSubfieldWithCode('a'));
+
+    // 0. Check that requirements for all records, independent of type or journal are met:
+    CheckGenericRequirements(record, reasons_for_being_invalid);
+
+    // 1. Check that present fields meet all the requirements:
+    MARC::Tag last_tag("   ");
+    const auto &validators(record.isReviewArticle() ? review_article_validators : regular_article_validators);
+    std::set<std::string> present_tags;
+    for (const auto &field : record) {
+        const auto current_tag(field.getTag());
+        if (current_tag == last_tag and not field.isRepeatableField())
+            reasons_for_being_invalid->append(current_tag.toString() + " is not a repeatable field\n");
+        last_tag = current_tag;
+        present_tags.emplace(current_tag.toString());
+
+        for (const auto validator : validators) {
+            std::string reason_for_being_invalid;
+            if (validator->foundRuleMatch(ZID, field, &reason_for_being_invalid)) {
+                if (not reason_for_being_invalid.empty())
+                    reasons_for_being_invalid->append(reason_for_being_invalid + "\n");
+                break;
+            }
+        }
     }
 
-    if (not RecordMeetsExpectations(record, journal_name_and_info->first, general_info, journal_name_and_info->second, error_message))
-        return false;
-    else if (not journal_name_and_info->second.isInDatabase()) {
-        AnalyseNewJournalRecord(record, first_record, general_info, &journal_name_and_info->second);
-        ++(*new_record_count);
-    }
+    // 2. Check for missing required fields:
+    std::set<std::string> missing_tags;
+    for (const auto validator : validators)
+        validator->findMissingTags(ZID, present_tags, &missing_tags);
+    for (const auto &missing_tag : missing_tags)
+        reasons_for_being_invalid->append("required " + missing_tag + "-field is missing\n");
 
-    return true;
+    return reasons_for_being_invalid->empty();
 }
 
 
 } // unnamed namespace
 
 
 int Main(int argc, char *argv[]) {
-    if (argc != 5 and argc != 6)
-        Usage();
+    if (argc != 5)
+        ::Usage("marc_input marc_output missed_expectations_file email_address");
```

Comment:
It's OK if you remove the update_db mode, but what happened to the [--update-db-errors] flag for the regular mode?

Reply:
Done